File: medbrains/crates/medbrains-core/src/specialty/cath_lab.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "cath_procedure_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum CathProcedureType {
    DiagnosticCath,
    Pci,
    Pacemaker,
    Icd,
    Eps,
    Ablation,
    ValveIntervention,
    Structural,
    Peripheral,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "stemi_pathway_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum StemiPathwayStatus {
    Door,
    Ecg,
    CathLabActivation,
    ArterialAccess,
    BalloonInflation,
    Completed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "hemodynamic_site", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum HemodynamicSite {
    Aorta,
    Lv,
    Rv,
    Ra,
    La,
    Pa,
    Pcwp,
    Svg,
    Lm,
    Lad,
    Lcx,
    Rca,
    Other,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "cath_device_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum CathDeviceType {
    Stent,
    Balloon,
    Guidewire,
    Catheter,
    ClosureDevice,
    Pacemaker,
    Icd,
    Lead,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CathProcedure {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub patient_id: Uuid,
    pub procedure_type: CathProcedureType,
    pub operator_id: Uuid,
    pub assistant_ids: Vec<Uuid>,
    pub is_stemi: bool,
    pub door_time: Option<DateTime<Utc>>,
    pub balloon_time: Option<DateTime<Utc>>,
    pub door_to_balloon_minutes: Option<i32>,
    pub fluoroscopy_time_seconds: Option<i32>,
    pub total_dap: Option<rust_decimal::Decimal>,
    pub total_air_kerma: Option<rust_decimal::Decimal>,
    pub contrast_type: Option<String>,
    pub contrast_volume_ml: Option<rust_decimal::Decimal>,
    pub access_site: Option<String>,
    pub findings: serde_json::Value,
    pub complications: Option<String>,
    pub status: String,
    pub scheduled_at: Option<DateTime<Utc>>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CathHemodynamic {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub procedure_id: Uuid,
    pub site: HemodynamicSite,
    pub systolic_mmhg: Option<rust_decimal::Decimal>,
    pub diastolic_mmhg: Option<rust_decimal::Decimal>,
    pub mean_mmhg: Option<rust_decimal::Decimal>,
    pub saturation_pct: Option<rust_decimal::Decimal>,
    pub gradient_mmhg: Option<rust_decimal::Decimal>,
    pub recorded_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CathDevice {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub procedure_id: Uuid,
    pub device_type: CathDeviceType,
    pub manufacturer: Option<String>,
    pub model: Option<String>,
    pub lot_number: Option<String>,
    pub barcode: Option<String>,
    pub size: Option<String>,
    pub is_consignment: bool,
    pub vendor_id: Option<Uuid>,
    pub unit_cost: Option<rust_decimal::Decimal>,
    pub billed: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CathStemiTimeline {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub procedure_id: Uuid,
    pub event: StemiPathwayStatus,
    pub event_time: DateTime<Utc>,
    pub recorded_by: Uuid,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CathPostMonitoring {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub procedure_id: Uuid,
    pub monitored_at: DateTime<Utc>,
    pub sheath_status: Option<String>,
    pub access_site_status: Option<String>,
    pub vitals: serde_json::Value,
    pub ambulation_started: bool,
    pub monitored_by: Uuid,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: medbrains/packages/types/src/cath-lab.ts
// Cath lab types — split from index.ts, barrel-re-exported.

// ── Cath Lab ──

export type CathProcedureType =
  | "diagnostic_cath"
  | "pci"
  | "pacemaker"
  | "icd"
  | "eps"
  | "ablation"
  | "valve_intervention"
  | "structural"
  | "peripheral";

export type StemiPathwayStatus =
  | "door"
  | "ecg"
  | "cath_lab_activation"
  | "arterial_access"
  | "balloon_inflation"
  | "completed";

export type HemodynamicSite =
  | "aorta"
  | "lv"
  | "rv"
  | "ra"
  | "la"
  | "pa"
  | "pcwp"
  | "svg"
  | "lm"
  | "lad"
  | "lcx"
  | "rca"
  | "other";

export type CathDeviceType =
  | "stent"
  | "balloon"
  | "guidewire"
  | "catheter"
  | "closure_device"
  | "pacemaker"
  | "icd"
  | "lead"
  | "other";

/** Ophthalmology — a comprehensive eye exam. Per-eye: OD = right, OS = left. */
export interface OphthoExam {
  id: string;
  tenant_id: string;
  patient_id: string;
  encounter_id: string | null;
  visual_acuity_od: string | null;
  visual_acuity_os: string | null;
  sphere_od: string | null;
  sphere_os: string | null;
  cylinder_od: string | null;
  cylinder_os: string | null;
  axis_od: number | null;
  axis_os: number | null;
  iop_od: string | null;
  iop_os: string | null;
  slit_lamp: string | null;
  fundus: string | null;
  diagnosis: string | null;
  plan: string | null;
  examined_by: string | null;
  examined_at: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface CreateOphthoExamRequest {
  patient_id: string;
  encounter_id?: string | null;
  visual_acuity_od?: string | null;
  visual_acuity_os?: string | null;
  sphere_od?: string | null;
  sphere_os?: string | null;
  cylinder_od?: string | null;
  cylinder_os?: string | null;
  axis_od?: number | null;
  axis_os?: number | null;
  iop_od?: string | null;
  iop_os?: string | null;
  slit_lamp?: string | null;
  fundus?: string | null;
  diagnosis?: string | null;
  plan?: string | null;
  status?: string | null;
}

export interface UpdateOphthoExamRequest {
  slit_lamp?: string | null;
  fundus?: string | null;
  diagnosis?: string | null;
  plan?: string | null;
  status?: string | null;
}

/** Dental — a dental exam (visit) + its tooth-wise chart entries. */
export interface DentalExam {
  id: string;
  tenant_id: string;
  patient_id: string;
  encounter_id: string | null;
  chief_complaint: string | null;
  diagnosis: string | null;
  treatment_plan: string | null;
  examined_by: string | null;
  examined_at: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface DentalChartEntry {
  id: string;
  tenant_id: string;
  exam_id: string;
  tooth_number: string;
  condition: string;
  surface: string | null;
  treatment_planned: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateDentalExamRequest {
  patient_id: string;
  encounter_id?: string | null;
  chief_complaint?: string | null;
  diagnosis?: string | null;
  treatment_plan?: string | null;
  status?: string | null;
}

export interface CreateDentalChartEntryRequest {
  tooth_number: string;
  condition: string;
  surface?: string | null;
  treatment_planned?: string | null;
  notes?: string | null;
}

/** Oncology depth — cancer TNM staging + radiation-therapy sessions. */
export interface CancerStaging {
  id: string;
  tenant_id: string;
  patient_id: string;
  primary_site: string;
  histology: string | null;
  t_stage: string | null;
  n_stage: string | null;
  m_stage: string | null;
  overall_stage: string | null;
  staged_by: string | null;
  staged_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface RadiationSession {
  id: string;
  tenant_id: string;
  patient_id: string;
  site: string;
  technique: string | null;
  total_dose_gy: string | null;
  fractions: number | null;
  session_number: number | null;
  delivered_by: string | null;
  delivered_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateCancerStagingRequest {
  patient_id: string;
  primary_site: string;
  histology?: string | null;
  t_stage?: string | null;
  n_stage?: string | null;
  m_stage?: string | null;
  overall_stage?: string | null;
  notes?: string | null;
}

export interface CreateRadiationRequest {
  patient_id: string;
  site: string;
  technique?: string | null;
  total_dose_gy?: string | null;
  fractions?: number | null;
  session_number?: number | null;
  notes?: string | null;
}

export interface CathProcedure {
  id: string;
  tenant_id: string;
  patient_id: string;
  procedure_type: CathProcedureType;
  operator_id: string;
  is_stemi: boolean;
  door_time: string | null;
  balloon_time: string | null;
  door_to_balloon_minutes: number | null;
  fluoroscopy_time_seconds: number | null;
  total_dap: number | null;
  total_air_kerma: number | null;
  contrast_type: string | null;
  contrast_volume_ml: number | null;
  findings: Record<string, unknown> | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface CreateCathProcedureRequest {
  patient_id: string;
  procedure_type: CathProcedureType;
  operator_id: string;
  is_stemi?: boolean;
  door_time?: string;
  contrast_type?: string;
  contrast_volume_ml?: number;
  findings?: Record<string, unknown>;
}

export interface CathHemodynamic {
  id: string;
  tenant_id: string;
  procedure_id: string;
  site: HemodynamicSite;
  systolic_mmhg: number | null;
  diastolic_mmhg: number | null;
  mean_mmhg: number | null;
  saturation_pct: number | null;
  gradient_mmhg: number | null;
  created_at: string;
}

export interface CreateCathHemodynamicRequest {
  site: HemodynamicSite;
  systolic_mmhg?: number;
  diastolic_mmhg?: number;
  mean_mmhg?: number;
  saturation_pct?: number;
  gradient_mmhg?: number;
}

export interface CathDevice {
  id: string;
  tenant_id: string;
  procedure_id: string;
  device_type: CathDeviceType;
  manufacturer: string | null;
  lot_number: string | null;
  barcode: string | null;
  is_consignment: boolean;
  vendor_id: string | null;
  unit_cost: number | null;
  billed: boolean;
  created_at: string;
}

export interface CreateCathDeviceRequest {
  device_type: CathDeviceType;
  manufacturer?: string;
  lot_number?: string;
  barcode?: string;
  is_consignment?: boolean;
  vendor_id?: string;
  unit_cost?: number;
}

export interface CathStemiTimeline {
  id: string;
  tenant_id: string;
  procedure_id: string;
  event: StemiPathwayStatus;
  event_time: string;
  recorded_by: string;
  created_at: string;
}

export interface CreateCathStemiEventRequest {
  event: StemiPathwayStatus;
  event_time: string;
}

export interface CathPostMonitoring {
  id: string;
  tenant_id: string;
  procedure_id: string;
  monitored_at: string;
  sheath_status: string | null;
  access_site_status: string | null;
  vitals: Record<string, unknown> | null;
  ambulation_started: boolean;
  created_at: string;
}

export interface CreateCathPostMonitoringRequest {
  monitored_at: string;
  sheath_status?: string;
  access_site_status?: string;
  vitals?: Record<string, unknown>;
  ambulation_started?: boolean;
}
